_posts/2016-06-27-installing-xhprof.md

```diff
@@ -0,0 +1,138 @@
+
+I find the easiest way to do this is just to symlink the `xhprof_html` directory into the existing webroot of whatever site you're working on locally, for example:
+
+`ln -s /usr/share/php/xhprof_html /var/www/my-website-dir/xhprof`
+
+If you're using DrupalVM, a separate vhost configuration will already be set up for XHProf, and the default URL is `http://xhprof.drupalvm.dev/` although it can be changed in your `config.yml` file.
+
+### Hooking XHProf into your module code
+
+Generally, the process of profiling a chunk of code using XHProf goes as follows:
+
+1. Call `xhprof_enable()`
+2. Run the code you want profiled
+3. Once the code has finished running, call `xhprof_disable()`. That function will return the profiler data, which you can either display to the screen (not recommended), or...
+4. Store the profiler data to a file by creating a new `XHProfRuns_Default();` object and calling its `save_run` method.
+
+In this case, I've been profiling a module that implements a few drush commands from the command line which I'd like to optimize. So I created `_modulename_xhprof_enable()` and `_modulename_xhprof_disable()` functions -- the names don't matter here -- and then added a `--profile` flag to my drush command options which, when it is set to true, calls my custom enable/disable functinos before and after the drush command runs.
+
+Here's what those look like in full:
+
+```php
+/**
+ * Helper function to enable xhprof.
+ */
+function _mymodule_enable_xhprof() {
+  if (function_exists('xhprof_enable')) {
+    // Tell XHProf to track both CPU time and memory usage
+    xhprof_enable(XHPROF_FLAGS_CPU + XHPROF_FLAGS_MEMORY,
+      array(
+        // Don't treat these functions as separate function callss
+        // in the results.
+        'ignored_functions' => array('call_user_func',
+          'call_user_func_array',
+        ),
+      ));
+  }
+}
+
+/**
+ * Helper function to disable xhprof and save logs.
+ */
+function _mymodule_disable_xhprof() {
+  if (function_exists('xhprof_enable')) {
+    $xhprof_data = xhprof_disable();
+
+    //
+    // Saving the XHProf run
+    // using the default implementation of iXHProfRuns.
+    //
+    include_once "/usr/share/php/xhprof_lib/utils/xhprof_lib.php";
+    include_once "/usr/share/php/xhprof_lib/utils/xhprof_runs.php";
+
+    $xhprof_runs = new XHProfRuns_Default();
+
+    // Save the run under a namespace "xhprof_foo".
+    //
+    // **NOTE**:
+    // By default save_run() will automatically generate a unique
+    // run id for you. [You can override that behavior by passing
+    // a run id (optional arg) to the save_run() method instead.]
+    // .
+    $run_id = $xhprof_runs->save_run($xhprof_data, 'xhprof_mymodule');
+
+    echo "---------------\nAssuming you have set up the http based UI for \nXHProf at some address, you can view run at \nhttp://mywebsiteurl.dev/xhprof/index.php?run=$run_id&source=xhprof_mymodule\n---------------\n";
```

Comment:
Since this line is so long, it's appears chopped up on the editor on the blog. In other words, it's not clear to scroll right, so maybe just shorten it or make it multiple lines?